### ISSUE TITLE
Request: Callback to reduce lr on NaN loss

### ISSUE THREAD

**Issue by AndrasEros** [keras-team/keras#8332]
Feature request: Sorry if I'm raising this on the wrong place or in the wrong format.
I would like to ask for a callback that deals with the problem of the loss turning to NaN. The most common solution especially for large networks is to reduce the learning rate. On the other hand it's not ideal to reduce too much because the training will slow down. Currently we only have a callback that can be used to stop the training if the NaN problem happens. My best idea that it could work like this:

1. Backs up the weights at the beginning of each epoch
2. If during the epoch the loss turns to NaN then the epoch should be stopped
3. The weights should be restored with the backup from step 1
4. Learning rate should be reduced
5. Continue training with next epoch

Furthermore we could optionally increase the lr after the model starts converging. Having more enthusiasm than coding skills I wrote (copied a lot) this to be included in callbacks.py:

```python
class ReduceOnNaN(Callback):
    """Handle cases when loss turns to NaN by reducing lr.
    Backs up the weights in the beginning of each epoch
    and restores weights if NaN happens. After that
    the learning rate will be reduced. Changes can be
    reversed step by step if the model starts converging.
    # Arguments
        
        factor: factor by which the learning rate will
            be reduced. new_lr = lr * factor
        revert: start increasing lr after the model improves
    """

    def __init__(self, 
                 factor=0.25,
                 revert=False):
        super(ModelCheckpoint, self).__init__()
        self.factor = factor
        self.monitor = 'val_loss'
        self.revert = revert
        # Save initial learning rate
        self.initial_lr=float(K.get_value(self.model.optimizer.lr))
        # Define a temp file for saving the last set of weights
        self.path = "./temp/temp.h5"
        # Create the temp directory if it doesn't exists
        try:
            os.stat(directory)
        except:
            os.mkdir(directory) 

        # Reused auto monitor_op from ModelCheckpoint
        if 'acc' in self.monitor or self.monitor.startswith('fmeasure'):
            self.monitor_op = np.greater
            self.best = -np.Inf
        else:
            self.monitor_op = np.less
            self.best = np.Inf

    def on_epoch_begin(self, epoch, logs=None):
        #Save weights in the beginning of epoch
        self.model.save_weights(path, overwrite=True)

    def on_batch_end(self, batch, logs=None):
        logs = logs or {}
        loss = logs.get('loss')
        if loss is not None:
            # Check if we have NaN or Inf
            if np.isnan(loss) or np.isinf(loss):
                # We have the NaN problem
                # Load the weights from the begining of epoch
                model.load_weights(path)
                # Reduce the learning rate
                old_lr = float(K.get_value(self.model.optimizer.lr))
                new_lr = old_lr * self.factor
                K.set_value(self.model.optimizer.lr, new_lr)
                print('Reducing learning rate to %s.' % (new_lr))
                # Pass instruction to training.py to break the epoch
                self.model.stop_epoch = True
            elif (self.revert):
                # No NaN problem (anymore)
                current = logs.get(self.monitor)
                # Check if current epooch is better than any before
                if self.monitor_op(current, self.best):
                    self.best = current
                    old_lr = float(K.get_value(self.model.optimizer.lr))
                    # If learning rate was reduced compared to the initial one
                    if old_lr < self.initial_lr:
                        # Start increasing back with same speed as reduction
                        new_lr = old_lr / self.factor
                        K.set_value(self.model.optimizer.lr, new_lr)
                        print('Increasing learning rate to %s.' % (new_lr))
```
And this extension in training.py at line 1219:
```python
callbacks.on_batch_end(batch_index, batch_logs)
if callback_model.stop_training:
    break
if callback_model.stop_epoch:
    callback_model.stop_epoch = False
    break
```
My main argument for inclusion is that I consider Keras as a framework that saves time for the user to invest it in model design. This enhancement can save a lot of time in the manual experimenting process of finding the lr that is low enough but not too low if the loss runs to NaN. Secondly Keras could offer the common stackoverflow recommendation as a simply packaged solution.
*The posted code was not tested yet*
